internal/pkg/initialize/workload.go

```diff
@@ -0,0 +1,317 @@
+	AddJobToApp(app *config.Application, jobName string) error
+	AddServiceToApp(app *config.Application, serviceName string) error
+}
+
+// Workspace contains the methods needed to manipulate a Copilot workspace.
+type Workspace interface {
+	CopilotDirPath() (string, error)
+	WriteJobManifest(marshaler encoding.BinaryMarshaler, jobName string) (string, error)
+	WriteServiceManifest(marshaler encoding.BinaryMarshaler, serviceName string) (string, error)
+}
+
+// Prog contains the methods needed to render multi-stage operations.
+type Prog interface {
+	Start(label string)
+	Stop(label string)
+}
+
+// WorkloadProps contains the information needed to represent a Workload (job or service).
+type WorkloadProps struct {
+	App            string
+	Type           string
+	Name           string
+	DockerfilePath string
+	Image          string
+}
+
+// JobProps contains the information needed to represent a Job
+type JobProps struct {
+	*WorkloadProps
+	Schedule string
+	Timeout  string
+	Retries  int
+}
+
+// ServiceProps contains the information needed to represent a Service (port, HealthCheck, and workload common props)
+type ServiceProps struct {
+	*WorkloadProps
+	Port        uint16
+	HealthCheck *manifest.ContainerHealthCheck
+}
+
+// WorkloadInitializer holds the clients necessary to initialize either a
+// service or job in an existing application.
+type WorkloadInitializer struct {
+	Store    Store
+	Deployer WorkloadAdder
+	Ws       Workspace
+	Prog     Prog
+}
+
+// NewWorkloadInitializer returns a struct which holds the clients and configuration needed to
+// initialize a new workload in an existing application
+func NewWorkloadInitializer(s Store, ws Workspace, p Prog, d WorkloadAdder) *WorkloadInitializer {
+	return &WorkloadInitializer{
+		Store:    s,
+		Ws:       ws,
+		Prog:     p,
+		Deployer: d,
+	}
+}
+
+func (w *WorkloadInitializer) writeManifest(mf encoding.BinaryMarshaler, wlName string, wlType string) (string, error) {
+	switch wlType {
+	case svcWlType:
+		return w.Ws.WriteServiceManifest(mf, wlName)
+	case jobWlType:
+		return w.Ws.WriteJobManifest(mf, wlName)
+	default:
+		return "", fmt.Errorf(fmtErrUnrecognizedWlType, wlType)
+	}
+}
+
+func (w *WorkloadInitializer) addWlToApp(app *config.Application, wlName string, wlType string) error {
+	switch wlType {
+	case svcWlType:
+		return w.Deployer.AddServiceToApp(app, wlName)
+	case jobWlType:
+		return w.Deployer.AddJobToApp(app, wlName)
+	default:
+		return fmt.Errorf(fmtErrUnrecognizedWlType, wlType)
+	}
+}
+
+func (w *WorkloadInitializer) addWlToStore(wl *config.Workload, wlType string) error {
+	switch wlType {
+	case svcWlType:
+		return w.Store.CreateService(wl)
+	case jobWlType:
+		return w.Store.CreateJob(wl)
+	default:
+		return fmt.Errorf(fmtErrUnrecognizedWlType, wlType)
+	}
+}
+
+func (w *WorkloadInitializer) initWorkload(props *WorkloadProps, wlType string, mf encoding.BinaryMarshaler, port uint16, sched string) (manifestPath string, err error) {
```

Comment:
What do you think of splitting into two separate methods: initJob and initSvc? And moving any common piece to small methods? 

This way this function won't accept arguments that might not make sense for the other workload types

Reply:
Done in most recent push; all these comments have been addressed. tysm for the thoughtful feedback